File: frontend/src/features/exam/format.ts
// Pure presentation helpers for the exam feature — no React.
import type { TFn } from "@/lib/i18n/types";

/** Format a millisecond countdown as H:MM:SS (or MM:SS under an hour). */
export function fmtCountdown(ms: number): string {
  const total = Math.max(0, Math.floor(ms / 1000));
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  const mm = String(m).padStart(2, "0");
  const ss = String(s).padStart(2, "0");
  if (h > 0) return `${h}:${mm}:${ss}`;
  return `${mm}:${ss}`;
}

export function readinessLabel(t: TFn, level: string | null): string {
  if (!level) return "—";
  const k = `readiness.${level}`;
  const v = t(k);
  // Unknown levels fall back to a lowercased space-joined rendering of the key.
  return v === k ? level.replace(/_/g, " ") : v;
}

/** True when the remaining time is in the final warning window (<= 5 min). */
export function isTimeCritical(ms: number): boolean {
  return ms <= 5 * 60 * 1000;
}

File: mobile/lib/features/exam/format.dart
// Pure presentation helpers for exam (and shared) timing / percentages.

/// Format a millisecond countdown as `H:MM:SS` (or `MM:SS` under an hour).
String fmtCountdown(int ms) {
  final total = (ms < 0 ? 0 : ms) ~/ 1000;
  final h = total ~/ 3600;
  final m = (total % 3600) ~/ 60;
  final s = total % 60;
  final mm = m.toString().padLeft(2, '0');
  final ss = s.toString().padLeft(2, '0');
  if (h > 0) return '$h:$mm:$ss';
  return '$mm:$ss';
}

/// True when remaining time is in the final warning window (≤ 5 min).
bool isTimeCritical(int ms) => ms <= 5 * 60 * 1000;

/// Format a 0–1 accuracy ratio as a percent string.
String fmtPct(double n) => '${(n * 100).round()}%';

/// Human duration from milliseconds (`45s`, `1m 30s`, `1h 0m`).
String fmtDuration(int ms) {
  final totalSec = (ms / 1000).round();
  final h = totalSec ~/ 3600;
  final m = (totalSec % 3600) ~/ 60;
  final s = totalSec % 60;
  if (h > 0) return '${h}h ${m}m';
  if (m > 0) return '${m}m ${s}s';
  return '${s}s';
}
